### PR TITLE
Add multiple endpoint support

### PR DESCRIPTION
**Pull request by igable**
Add an Endpoint class as a part of the resource module. An Endpoint can be
optionally passed to any of the resource objects to allow them to
access a different DCM endpoint then the one specified either by
environment variables or in configuration files as loaded by the config
class.

- Issue #210 heavily inspired this feature, thanks @thijs-creemers
- Fixes Issue #207

### CODE REVIEW

**Comment by igable** [comment]
These changes need more testing. However, I would appreciate feedback on the direction from @zomGreg, @bdwilliams, @buzztroll, @thijs-creemers on the direction of this.

Here's few examples of how you would use this class:
```python
    saas_endpoint = Endpoint(url="https://dcm.enstratius.com/api/enstratus/2015-05-25",
                            api_version="2015-05-25",
                            access_key="abcdefg",
                            secret_key="gfedcba")

    vagrant_endpoint = Endpoint(url="https://vagrant.vm/api/enstratus/2015-05-25",
                                api_version="2015-05-25",
                                access_key="abcdefg",
                                secret_key="gfedcba",
                                ssl_verify=False)

    # load endpoints from json files
    from_file_endpoint = Endpoint.from_file("~/endpoints/endpoint.json")
    multiple_from_file =  Endpoint.multiple_from_file("~/endpoints/endpoints.json")


    # loop over a returned dictionary of endpoints
    for nickname, endpoint in multiple_from_file.items():
        print "Endpoint Nickname: %s" % nickname
        print "Endpoint url: %s" % endpoint.url
        print "Endpoint API Version: %s" % endpoint.api_version

    servers = Server.all(endpoint=saas_endpoint)

    r = Resource('infrastructure/Server',endpoint=from_file_endpoint)
```

As you can see above, you can load endpoint definitions from files that looks like: [endpoint.json](https://github.com/igable/mixcoatl/blob/feature/multi-endpoint/tests/data/unit/admin/endpoint.json) and  [endpoints.json](https://github.com/igable/mixcoatl/blob/feature/multi-endpoint/tests/data/unit/admin/endpoints.json)

File: mixcoatl/resource.py [hunk 1]
```diff
@@ -8,7 +8,89 @@
 from mixcoatl.settings.load_settings import settings
 from mixcoatl.decorators.lazy import lazy_property
 from mixcoatl.utils import camel_keys
+import json
 
+class Endpoint(object):
+    """ A class for representing DCM endpoints and loading them from json files
+    """
+
+    def __init__(self, url = None, basepath=None, api_version=None, secret_key=None, access_key=None, ssl_verify=True, nickname=None):
```
Comment (buzztroll) [inline mixcoatl/resource.py]: nit: delete the spaces: url=None (sorry, I am not happy that I was compelled to make this comment).

Comment (igable) [inline mixcoatl/resource.py]: On the contrary, I'm thrilled that I'm not the only one who is going to be making PEP8 related comments. Hit me with these anytime. A PEP8 commit is pending shortly.

**Comment by igable** [comment]
Talking with Brian notes:

- in the event that no API version is specified it should be parsed out of the URL. If it's not valid date if should fail.
- Need jenkins tests prior to merge of this pull request.

**Comment by buzztroll** [comment]
This looks like the right refactor.  I agree that it is better than using the config object only. +1

**Comment by igable** [comment]
This is now rebased after the 0.10.50.1 release and has testing against live DCM endpoints for server and volume create operations included in b938d33 . This PR, although it hits a lot files, primarily affects the Resource base class. Most of the changes outside of that look something like this:

```diff
class SubResource(Resource):
-    def __init__(self, some_id=None, *args, **kwargs):
-        Resource.__init__(self)
+    def __init__(self, some_id=None, endpoint=None, *args, **kwargs):
+        Resource.__init__(self, endpoint=endpoint)
```

Which makes me less nervous to merge this.

__

**Comment by igable** [comment]
To test this code in a virtualenv

```bash
mkdir mixcoatl
virtualenv mixcoatl
cd mixcoatl/
source bin/activate
git clone https://github.com/igable/mixcoatl.git
cd mixcoatl/
git fetch --all
git checkout feature/muti-endpoint

# install deps needed by nose testing and not regular deployment
pip install nose mock httpretty

# install
python setup.py develop

# go back to the top level virtualenv directory
cd..
```

Make the data for testing against a live DCM instance:

```bash
curl https://gist.githubusercontent.com/igable/73bf1c650553f9336c86/raw/758e084043efc3c4c6df3797ff474663b37007a2/jenkins.sh | bash

# replace the access and secret keys
vim secret-test-data/endpoints.json
```



Run the nose tests that will create a vm and volune with names like mixcoatl-qa-abcdefgh:

```bash
cd mixcoatl/tests/live/

# run the nose tests for a server
nosetests --nocapture test_server.py

# run the nose test for a volume
nosetests --nocapture test_volume.py
```

If you want to make your own code using this look at the example above this in the PR comment.

**Comment by igable** [comment]
After quick chat with Greg I'm merging.

**Pull request merged**